File: app/src/main/java/ru/melod1n/vk/common/FragmentSwitcher.java
package ru.melod1n.vk.common;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import java.util.ArrayList;

import ru.melod1n.vk.R;
import ru.melod1n.vk.current.BaseFragment;
import ru.melod1n.vk.fragment.FragmentMessages;

public class FragmentSwitcher {

    private static Fragment currentFragment = null;
    private static ArrayList<Fragment> fragments = new ArrayList<>();

    public static FragmentMessages fragmentMessages = new FragmentMessages();

    public static void switchFragment(AppCompatActivity parentActivity, @Nullable Fragment parentFragment, @NonNull Fragment fragment, @Nullable Bundle extraData, boolean withBackStack) {
        if (currentFragment != null && fragment.getClass().getSimpleName().equals(currentFragment.getClass().getSimpleName())) {
            return;
        }

        if (extraData != null) fragment.setArguments(extraData);

        boolean[] containsList = new boolean[]{false, false};

        for (Fragment f : fragments) {
            if (f.getClass().getSimpleName().equals(fragment.getClass().getSimpleName())) {
                containsList[0] = true;
                break;
            }
        }

        boolean fromFragment = parentFragment != null;

        FragmentManager manager = fromFragment ? parentFragment.requireFragmentManager() : parentActivity.getSupportFragmentManager();

        for (Fragment f : manager.getFragments()) {
            if (f.getClass().getSimpleName().equals(fragment.getClass().getSimpleName())) {
                containsList[1] = true;
                break;
            }
        }

        boolean contains = containsList[0] && containsList[1];

        if (contains) {
            if (fragment instanceof BaseFragment) {
                ((BaseFragment) fragment).onReopen(extraData);
            }
        } else {
            if (!containsList[0])
                fragments.add(fragment);
        }

        FragmentTransaction transaction = manager.beginTransaction();

        if (currentFragment != null) {
            Fragment managerCurrentFragment = null;

            for (Fragment f : manager.getFragments()) {
                if (f.isVisible()) {
                    managerCurrentFragment = f;
                    break;
                }
            }

            if (managerCurrentFragment != currentFragment) {
                currentFragment = managerCurrentFragment;
            }

            if (currentFragment != null) {
                transaction.hide(currentFragment);
            }

            if (contains) {
                transaction.show(fragment);
            } else {
                transaction.add(R.id.fragment_container, fragment, fragment.getClass().getSimpleName());
            }
        } else {
            transaction.add(R.id.fragment_container, fragment, fragment.getClass().getSimpleName());
        }

        if (withBackStack) transaction.addToBackStack(fragment.getClass().getSimpleName());
        manager.addOnBackStackChangedListener(() -> {
            for (Fragment f : manager.getFragments()) {
                if (f.isVisible()) {
                    currentFragment = f;
                    break;
                }
            }
        });

        transaction.commit();

        currentFragment = fragment;
    }

    public static void switchFragment(AppCompatActivity parentActivity, @NonNull Fragment fragment) {
        switchFragment(parentActivity, null, fragment, null, false);
    }

}

File: app/src/main/java/ru/melod1n/vk/common/FragmentSwitcher.kt
package ru.melod1n.vk.common

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import ru.melod1n.vk.R
import ru.melod1n.vk.current.BaseFragment
import ru.melod1n.vk.util.ArrayUtil.isEmpty
import java.util.*

class FragmentSwitcher {
    fun switchFragment(parentActivity: AppCompatActivity, fragment: Fragment) {
        switchFragment(parentActivity, null, fragment, null, false)
    }

    companion object {
        private var currentFragmentTag: String? = null
        private val tags = ArrayList<String>()
        fun switchFragment(parentActivity: AppCompatActivity, parentFragment: Fragment?, fragment: Fragment, extraData: Bundle?, withBackStack: Boolean) {
            val fromFragment = parentFragment != null
            val manager = if (fromFragment) parentFragment!!.requireFragmentManager() else parentActivity.supportFragmentManager
            var currentFragment = getCurrentFragment(parentActivity.supportFragmentManager)
            if (currentFragment != null) {
                currentFragmentTag = currentFragment.javaClass.simpleName
                if (fragment.javaClass.simpleName == currentFragmentTag) {
                    if (!currentFragment.isAdded) {
                        manager.beginTransaction().add(R.id.fragment_container, currentFragment, currentFragment.javaClass.simpleName).commit()
                    }
                    return
                }
            }
            if (extraData != null) fragment.arguments = extraData
            val containsList = booleanArrayOf(false, false)
            for (tag in tags) {
                if (tag == fragment.javaClass.simpleName) {
                    containsList[0] = true
                    break
                }
            }
            for (f in manager.fragments) {
                if (f.javaClass.simpleName == fragment.javaClass.simpleName) {
                    containsList[1] = true
                    break
                }
            }
            val contains = containsList[0] && containsList[1]
            if (contains) {
                if (fragment is BaseFragment) {
                    fragment.requestReopen(extraData)
                }
            } else {
                if (!containsList[0]) tags.add(fragment.javaClass.simpleName)
            }
            val transaction = manager.beginTransaction()
            if (currentFragment != null) {
                var managerCurrentFragment: Fragment? = null
                for (f in manager.fragments) {
                    if (f.isVisible) {
                        managerCurrentFragment = f
                        break
                    }
                }
                if (managerCurrentFragment !== currentFragment) {
                    currentFragment = managerCurrentFragment
                }
                if (currentFragment != null) {
                    transaction.hide(currentFragment)
                }
                if (contains) {
                    transaction.show(fragment)
                } else {
                    transaction.add(R.id.fragment_container, fragment, fragment.javaClass.simpleName)
                }
            } else {
                transaction.add(R.id.fragment_container, fragment, fragment.javaClass.simpleName)
            }
            if (withBackStack) transaction.addToBackStack(fragment.javaClass.simpleName)
            manager.addOnBackStackChangedListener {
                for (f in manager.fragments) {
                    if (f.isVisible) {
                        currentFragmentTag = f.javaClass.simpleName
                        break
                    }
                }
            }
            transaction.commit()
            currentFragmentTag = fragment.javaClass.simpleName
        }

        fun getCurrentFragment(fragmentManager: FragmentManager): Fragment? {
            val fragments = fragmentManager.fragments
            if (isEmpty(fragments)) return null
            for (fragment in fragments) {
                if (fragment!!.isVisible) return fragment
            }
            return null
        }

        val instance: FragmentSwitcher
            get() = FragmentSwitcher()
    }
}